python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -14,13 +17,337 @@
 logger = logging.getLogger(__name__)
 
 
-class _DistributeResources:
-    """Generic functionality for resource allocation functions"""
+@PublicAPI(stability="beta")
+class DistributeResources:
+    """This class creates a basic uniform resource allocation function.
 
-    def __init__(self, add_bundles: bool = False):
-        """If add_bundles is True, create new bundles from free resources.
-        Otherwise, spread them among base_trial_resource bundles."""
+    The function naively balances free resources (CPUs and GPUs) between
+    trials, giving them all equal priority, ensuring that all resources
+    are always being used. The free resources will be placed in new bundles.
+    The function assumes that all bundles are equal (there is no "head"
+    bundle).
+
+    If for some reason a trial ends up with
+    more resources than there are free ones, it will adjust downwards.
```

Comment:
why would they be in such situation? Is that the cluster is also being autoscaled at the same time?

Reply:
Yes, this is a measure to ensure training does not stall if nodes go down.